common/core/overlaylayer.cpp

```diff
@@ -229,55 +229,45 @@ void OverlayLayer::InitializeState(HwcLayer* layer,
 
   int32_t left_constraint = layer->GetLeftConstraint();
   int32_t right_constraint = layer->GetRightConstraint();
-  if (left_constraint >= 0 && right_constraint >= 0) {
-    if (display_frame_.right > right_constraint) {
-      display_frame_.right = right_constraint;
-    }
+  int32_t left_source_constraint = layer->GetLeftSourceConstraint();
+  int32_t right_source_constraint = layer->GetRightSourceConstraint();
+  int32_t display_frame_left = display_frame_.left;
+  uint32_t frame_width = display_frame_.right - display_frame_.left;
+  uint32_t source_width = source_crop_.right - source_crop_.left;
+  uint32_t frame_offset_left = 0;
+  uint32_t frame_offset_right = frame_width;
 
-    if (display_frame_.left < left_constraint) {
-      display_frame_.left = left_constraint;
-    }
-
-    if (display_frame_.right < right_constraint) {
-      display_frame_.right =
-          std::max(display_frame_.left, display_frame_.right);
-    }
+  if (left_constraint >= 0 && right_constraint >= 0) {
+    if (display_frame_.right < left_source_constraint ||
```

Comment:
hmm not sure whhy source constaint,s  should impact this. I think the only change we should probably do from original code is as follows perhaps :

if (display_frame_.left < left_constraint) {
  display_frame_.left =  0

Reply:
but than line 241 in original code should handle the translation of coordinates. Wondering what this new change is fixing

Reply:
as the display frame rect need *source_constraint to identify display index which be used currently. with only *_constraint, we don't know the offset of the display frame for second,3rd... displays, and the display rect will use wrong left/right value.

Reply:
i added the original error log here to play 3840x1080 video on frame rect 960x2880(middle of the final screen) in two 1920x1080 displays mosaic mode, the original code will only show:
`11-14 03:30:59.190  2472 16269 I iahwcomposer: Display index 1
11-14 03:30:59.190  2472 16269 I iahwcomposer: dlconstraint 0
11-14 03:30:59.190  2472 16269 I iahwcomposer: drconstraint 1920
11-14 03:30:59.190  2472 16269 I iahwcomposer: right_constraint 3840
11-14 03:30:59.190  2472 16269 I iahwcomposer: left_constraint 1920
11-14 03:30:59.190  2472 16269 I iahwcomposer: ======= Yugang got layer =======
11-14 03:30:59.190  2472 16269 I iahwcomposer: ======================= Yugang no scaling ===========================
11-14 03:30:59.190  2472 16269 I iahwcomposer: before: lc=0, rc=1920, lsc=1920, rsc=3840
11-14 03:30:59.190  2472 16269 I iahwcomposer:         sc_l=0.000000, sc_r=3840.000000, df_l=960, df_r=2880
11-14 03:30:59.190  2472 16269 I iahwcomposer: after: sc_l=1920.000000, sc_r=3840.000000, df_l=960, df_r=1920
11-14 03:30:59.190  2472 16269 I iahwcomposer: =====================================================================`

you can see the second display use the wrong display rect `df_l=960, df_r=1920`, should be "df_l=0, df_r=960"

Reply:
and the original code also can't split the part of the video to second, 3rd... displays in mosaic mode. the new change also fixed it

Reply:
and i just find the mouse in the second display of the mosaic mode also works after this change, but not sure why

Reply:
that means something is going wrong at line 250 in original code. We should be translating the coordinates of display frame rect . Something is wrong at display_frame_.right = right_constraint - display_frame_.right;  I think this one should be:

display_frame_.right = display_frame_.right - left_constraint;

Reply:
also check if left_constraint is being set correctly and greater than 0

Reply:
Can you check with this branch: https://github.com/intel/IA-Hardware-Composer/commits/mosaic Did couple of fixes  :   
1) Ignore layer if it is not part of this display.
2) Fix displayframe right calculation.

Reply:
@kalyankondapally tried https://github.com/intel/IA-Hardware-Composer/commits/next latest on 1A:
1. to play 1280x800 or 3840x1080 mp4 videos, still show half screen(in the first screen) in two displays mosaic mode
2. right-bottom half cycle icon was not there
3. the second screen also has no cursor 

i didn't reproduced icon issue(the icons which resemble the seating position), not sure if you enabled the surfacedamage?

Reply:
For the cursor, do we for sure know it's co-ordinates are in the missing display ? Layers will not show if the display frame rect of that layer is not part of this display. Even for Video, that's what I would check.

Reply:
yes, clicking is normally in second screen.

Reply:
I mean can u dump the cursor coordinates.

Reply:
Also check why they are not getting translated here: https://github.com/intel/IA-Hardware-Composer/blob/master/common/core/overlaylayer.cpp#L251

Reply:
not sure how to identify the layer type of cursor, but i got the print log by the probable position of current cursor(i add the log before original line 232 and after the original line 285):
`
11-11 15:13:04.619  2478  5397 I iahwcomposer: Display index 1

11-11 15:13:04.619  2478  5397 I iahwcomposer: dlconstraint 0

11-11 15:13:04.619  2478  5397 I iahwcomposer: drconstraint 1920

11-11 15:13:04.619  2478  5397 I iahwcomposer: right_constraint 3840

11-11 15:13:04.619  2478  5397 I iahwcomposer: left_constraint 1920

......................................................................

11-11 15:13:04.620  2478  5397 I iahwcomposer: Before=========================>

11-11 15:13:04.620  2478  5397 I iahwcomposer:  Display frame rect( r=3656, l=3634 ) Source crop( r=22.000000, l=0.000000 )

11-11 15:13:04.620  2478  5397 I iahwcomposer:  left_constraint=0, right_constraint=1920

11-11 15:13:04.620  2478  5397 I iahwcomposer:  left_source_constraint=1920, right_source_constraint=3840

11-11 15:13:04.620  2478  5397 I iahwcomposer: Before End =========================>

11-11 15:13:04.620  2478  5397 I iahwcomposer: After =========================>

11-11 15:13:04.620  2478  5397 I iahwcomposer:  Display frame rect( r=1920, l=1920 ) Source crop( r=1920.000000, l=1920.000000 )

11-11 15:13:04.620  2478  5397 I iahwcomposer:  left_constraint=0, right_constraint=1920

11-11 15:13:04.620  2478  5397 I iahwcomposer:  left_source_constraint=1920, right_source_constraint=3840

11-11 15:13:04.620  2478  5397 I iahwcomposer: After End =========================>

`

the Display frame rect is ( r=1920, l=1920 ), that is why nor show the cursor. and already been fixed in this PR, that is why the cursor can show normally after using this PR. for disappeared right-bottom icon issue, should have the same situation according the probable position of the icon.

so what is your concerns of this PR, is "hmm not sure whhy source constaint,s should impact this"? i think we have to use it to convert the rect for extended displays of mosaic mode.

Reply:
Source rect should be the region which should be interesting from which portion the buffer needs to be read from, displayframe will be where it needs to written to.

I see the issue now, display frame left position is always 0 as passed from Mosaic display. I dont see concern on this PR but wanted to better understand why the translation was not working.

Can u rebase this ontop of next branch and also remove this change:

  display_frame_.left > right_source_constraint) {
+      display_frame_.left = 0;
+      display_frame_.right = 0;
+      frame_offset_right = 0;

As in this case the layer should be marked as invisible and ignored.